File: app/src/main/java/ch/zuehlke/sbb/reddit/features/login/LoginFragment.java
package ch.zuehlke.sbb.reddit.features.login;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.TextInputEditText;
import android.support.v4.app.Fragment;
import android.support.v7.widget.AppCompatButton;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;

import com.google.common.base.Strings;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

import ch.zuehlke.sbb.reddit.R;
import ch.zuehlke.sbb.reddit.features.overview.OverviewActivity;

import static com.google.common.base.Preconditions.checkNotNull;

/**
 * Created by chsc on 08.11.17.
 */

public class LoginFragment extends Fragment implements LoginContract.View {

    private static final String EMAIL_PATTERN = "[a-zA-Z0-9._-]+@[a-z]+\\.+[a-z]+";

    private LoginContract.Presenter mPresenter;
    private ProgressBar mProgessBar;
    private AppCompatButton mLoginButton;
    private TextInputEditText mUsername;
    private TextInputEditText mPassword;

    public static LoginFragment newInstance() {
        return new LoginFragment();
    }

    @Override
    public void onResume() {
        super.onResume();
        mPresenter.start();
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.fragment_login, container, false);
        mProgessBar = root.findViewById(R.id.progressBar);
        mLoginButton = root.findViewById(R.id.loginButton);
        mUsername = root.findViewById(R.id.username);
        mPassword = root.findViewById(R.id.password);

        mLoginButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                mPresenter.login(mUsername.getText().toString(),mPassword.getText().toString());
            }
        });


        mUsername.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                // Do nothing
            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                // Do nothing
            }

            @Override
            public void afterTextChanged(Editable editable) {
                if(editable.length() > 0 && verifyIsEmail(editable.toString())){
                    mUsername.setError(null);
                }else{
                     mUsername.setError(getString(R.string.login_screen_invalid_email));
                }
            }
        });

        mPassword.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void afterTextChanged(Editable editable) {
                if(verifyPasswordLength(editable.toString())){
                    mPassword.setError(null);
                }else{
                    mPassword.setError(getString(R.string.login_screen_invalid_password_length));
                }
            }
        });



        return root;

    }

    @Override
    public void setPresenter(@NonNull LoginContract.Presenter presenter) {
        this.mPresenter = checkNotNull(presenter);
    }

    @Override
    public boolean isActive() {
        return isAdded();
    }

    @Override
    public void setLoadingIndicator(boolean isActive) {
            mProgessBar.setVisibility(isActive ? View.VISIBLE : View.GONE);
    }

    @Override
    public void showInvalidUsername() {
        mUsername.setError(getString(R.string.login_screen_invalid_username));
    }

    @Override
    public void showInvalidPassword() {
        mPassword.setError(getString(R.string.login_screen_invalid_password));
    }


    @Override
    public void showRedditNews() {
        Intent intent = new Intent(getContext(),OverviewActivity.class);
        startActivity(intent);
        getActivity().finish();
    }


    private boolean verifyPasswordLength(String password){
        return (!Strings.isNullOrEmpty(password) && password.length() >=6);
    }

    private boolean verifyIsEmail(String email){
        Matcher matcher = android.util.Patterns.EMAIL_ADDRESS.matcher(email);
        return matcher.matches();
    }
}

File: app/src/main/java/ch/zuehlke/sbb/reddit/features/login/LoginFragment.kt
package ch.zuehlke.sbb.reddit.features.login

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import ch.zuehlke.sbb.reddit.R
import ch.zuehlke.sbb.reddit.features.news.NewsActivity
import kotlinx.android.synthetic.main.fragment_login.*


/**
 * Created by chsc on 08.11.17.
 */

class LoginFragment : Fragment() {


    companion object {
        fun newInstance(): LoginFragment {
            return LoginFragment()
        }
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val root = inflater!!.inflate(R.layout.fragment_login, container, false)

        return root

    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        val loginViewModel = ViewModelProviders.of(activity).get(LoginViewModel::class.java)
        loginViewModel.viewState.observe(this, Observer { viewState: LoginViewModel.ViewState? -> handleViewState(viewState) })
        loginButton.setOnClickListener{loginViewModel.login(username.text.toString(), password.text.toString())}
        username.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                // Do nothing
            }

            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                // Do nothing
            }

            override fun afterTextChanged(editable: Editable) {
                if (editable.isNotEmpty() && loginViewModel.verifyIsEmail(editable.toString())) {
                    usernameLayout.error = null
                } else {
                    usernameLayout.error = getString(R.string.login_screen_invalid_email)
                }
            }
        })

        password.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {

            }

            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {

            }

            override fun afterTextChanged(editable: Editable) {
                if (loginViewModel.verifyPasswordLength(editable.toString())) {
                    passwordLayout.error = null
                } else {
                    passwordLayout.error = getString(R.string.login_screen_invalid_password_length)
                }
            }
        })
    }

    fun handleViewState(viewState: LoginViewModel.ViewState?){
        when(viewState){
            LoginViewModel.ViewState.LOADING -> progressBar.visibility = View.VISIBLE
            LoginViewModel.ViewState.NONE -> progressBar.visibility = View.GONE
            LoginViewModel.ViewState.INVALID_PASSWORD -> usernameLayout.error = getString(R.string.login_screen_invalid_username)
            LoginViewModel.ViewState.INVALID_USERNAME -> passwordLayout.error = getString(R.string.login_screen_invalid_password)
            LoginViewModel.ViewState.INVALID_CREDENTIALS -> {
                usernameLayout.error = getString(R.string.login_screen_invalid_username)
                passwordLayout.error = getString(R.string.login_screen_invalid_password)
            }
            LoginViewModel.ViewState.LOGGED_IN -> {
                progressBar.visibility = View.GONE
                val intent = Intent(context, NewsActivity::class.java)
                startActivity(intent)
                activity.finish()
            }
        }
    }


}
